### PR TITLE
feat: add network-parameterized block responses

### LINKED ISSUE #267: [Feature] Add network-parameterized block responses
### Component

network, json-rpc, providers, pubsub

### Describe the feature you would like

#190 replaced the provider with a network-parameterized provider, but block responses are still just Ethereum headers and Ethereum transactions. We need a `Block` type generic for networks.

We have one in `alloy-network` but we need to check that it is correct and use it in `alloy-providers`:

https://github.com/alloy-rs/alloy/blob/63af0c4dd43a94126348afaf94b744b98608c293/crates/network/src/lib.rs#L36-L51

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leruaa**
Closes #267

### CODE REVIEW

File: crates/network/src/lib.rs [hunk 1]
```diff
@@ -33,24 +34,22 @@ pub use any::AnyNetwork;
...
-/// A block response
-#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
-pub struct BlockResponse<N: Network> {
-    #[serde(flatten)]
-    header: N::HeaderResponse,
-    transactions: TransactionList<N::TransactionResponse>,
+/// A header response.
+pub trait HeaderResponse {
+    /// Base fee per unit of gas (if past London)
+    fn base_fee_per_gas(&self) -> &Option<U256>;
```
Comment (mattsse) [inline crates/network/src/lib.rs:48]:
```suggestion
    fn base_fee_per_gas(&self) -> Option<U256>;
```

File: crates/network/src/lib.rs [hunk 2]
```diff
@@ -33,24 +34,22 @@ pub use any::AnyNetwork;
...
+    /// Returns the blob fee for the next block according to the EIP-4844 spec.
+    ///
+    /// Returns `None` if `excess_blob_gas` is None.
+    fn next_block_blob_fee(&self) -> Option<u128>;
```
Comment (mattsse) [inline crates/network/src/lib.rs:52]: type mixing for fees is not ideal, perhaps we should just use U256 everywhere for fees  @prestwich

Comment (mattsse) [inline crates/network/src/lib.rs:52]: for header we need a few more things, but all of these should be moved to a header trait I believe

File: crates/network/src/lib.rs [hunk 3]
```diff
@@ -60,7 +59,7 @@ pub struct BlockResponse<N: Network> {
 /// [`TxReceipt`]: alloy_consensus::TxReceipt
 pub trait ReceiptResponse {
     /// Address of the created contract, or `None` if the transaction was not a deployment.
-    fn contract_address(&self) -> Option<Address>;
+    fn contract_address(&self) -> &Option<Address>;
```
Comment (mattsse) [inline crates/network/src/lib.rs:62]:
```suggestion
    fn contract_address(&self) -> Option<Address>;
```

File: crates/rpc-types/src/transaction_list.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+use alloy_primitives::B256;
+use serde::{Deserialize, Serialize};
+
+/// A list of transactions, either hydrated or hashes.
```
Comment (mattsse) [inline crates/rpc-types/src/transaction_list.rs:4]:
```suggestion
/// A list of transactions, either full, hashes or uncle for uncle blocks pre-merge.
```

File: crates/network/src/lib.rs [hunk 5]
```diff
@@ -33,24 +34,22 @@ pub use any::AnyNetwork;
...
-#[serde(untagged)]
-pub enum TransactionList<T> {
-    /// Hashes only.
-    Hashes(Vec<B256>),
-    /// Hydrated tx objects.
-    Hydrated(Vec<T>),
-    /// Special case for uncle response
-    Uncled,
+/// A block response.
+pub trait BlockResponse<N: Network> {
```
Comment (mattsse) [inline crates/network/src/lib.rs:38]: here we can also integrate a few more helper functions that currently exist on block, for iterating over txs for example.

**Review by mattsse — CHANGES_REQUESTED** [review]
a few suggestions,

it looks like we also need traits for header+block and enshrine Self::Header, Self::Block in the rpc response objects @prestwich ?

File: crates/rpc-types/src/transaction_list.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+use alloy_primitives::B256;
+use serde::{Deserialize, Serialize};
+
+/// A list of transactions, either full, hashes or uncle for uncle blocks pre-merge.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum TransactionList<T> {
```
Comment (prestwich) [inline crates/rpc-types/src/transaction_list.rs:7]: We can set a default type here

File: crates/rpc-types/src/eth/block.rs [hunk 7]
```diff
@@ -144,20 +138,7 @@ impl Header {
...
-    /// Only hashes
-    Hashes(Vec<B256>),
-    /// Full transactions
-    Full(Vec<Transaction>),
-    /// Special case for uncle response.
-    Uncle,
-}
-
-impl BlockTransactions {
+impl TransactionList<Transaction> {
```
Comment (prestwich) [inline crates/rpc-types/src/eth/block.rs:141]: could some of this be `impl<T: Transaction> TransactionList<T>` ?

**Review by prestwich — CHANGES_REQUESTED** [review]

File: crates/rpc-types/src/eth/block.rs (hunk 7 above)
Comment (prestwich) [reply crates/rpc-types/src/eth/block.rs:141]: for some `Transaction` trait 🤔

**Comment by prestwich** [comment]
looking pretty good. trying to figure out what's both maximally simple and maximally reusable for other network implementers

File: crates/network/src/lib.rs (hunk 2 above)
Comment (prestwich) [reply crates/network/src/lib.rs:52]: we need to action #433

**Comment by leruaa** [comment]
@prestwich I'm starting to get confused about what is left to be done in this PR, and if I understood the comments above correctly. In particular:

- Was I correct to rename `BlockResponse` and `HeaderResponse` traits to just `Block` and `Header`?
- I noticed there is a associated `Header` type on `Network` but it's used nowhere. Where is it supposed to be used?
- Should I add associated  `Block` and `Traansaction` types on `Network`?

Thanks in advance for your time

**Pull request closed without merging**